File: keytagger/scanner.py
import hashlib
import os
import time
from dataclasses import dataclass
from typing import Callable, Iterable, List, Optional, Tuple

from PIL import Image, ExifTags
import imagehash

from .db import Database

SUPPORTED_IMAGE_EXTENSIONS = {
    ".jpg", ".jpeg", ".png", ".webp", ".bmp", ".tif", ".tiff", ".gif"
}

SUPPORTED_VIDEO_EXTENSIONS = {
    ".mp4", ".mov", ".avi", ".mkv", ".webm", ".m4v", ".wmv", ".3gp"
}

SUPPORTED_AUDIO_EXTENSIONS = {
    ".m4a", ".mp3"
}


@dataclass
class ScanResult:
    scanned: int
    added_or_updated: int
    errors: int


def is_image_file(path: str) -> bool:
    _, ext = os.path.splitext(path)
    return ext.lower() in SUPPORTED_IMAGE_EXTENSIONS


def is_video_file(path: str) -> bool:
    _, ext = os.path.splitext(path)
    return ext.lower() in SUPPORTED_VIDEO_EXTENSIONS


def is_audio_file(path: str) -> bool:
    _, ext = os.path.splitext(path)
    return ext.lower() in SUPPORTED_AUDIO_EXTENSIONS


def sha256_of_file(path: str, chunk_size: int = 1024 * 1024) -> str:
    digest = hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(chunk_size), b""):
            digest.update(chunk)
    return digest.hexdigest()


def exif_capture_time_utc(img: Image.Image) -> Optional[int]:
    try:
        exif = img.getexif()
        if not exif:
            return None
        tags = {ExifTags.TAGS.get(k, k): v for k, v in exif.items()}
        dt = tags.get("DateTimeOriginal") or tags.get("DateTime")
        if not dt:
            return None
        # dt format: 'YYYY:MM:DD HH:MM:SS'
        try:
            import datetime as _dt

            local = _dt.datetime.strptime(str(dt), "%Y:%m:%d %H:%M:%S")
            epoch = int(local.timestamp())
            return epoch
        except Exception:
            return None
    except Exception:
        return None


def ensure_dir(path: str) -> None:
    os.makedirs(path, exist_ok=True)


def thumbnail_path_for(file_sha256: str, thumbnails_dir: str) -> str:
    ensure_dir(thumbnails_dir)
    return os.path.join(thumbnails_dir, f"{file_sha256}.jpg")


def save_thumbnail(img: Image.Image, thumb_path: str, max_size: int = 512) -> None:
    img_copy = img.copy()
    img_copy.thumbnail((max_size, max_size))
    img_copy = img_copy.convert("RGB")
    img_copy.save(thumb_path, format="JPEG", quality=85)


def create_video_thumbnail(video_path: str, thumb_path: str, max_size: int = 512) -> Optional[Tuple[int, int]]:
    try:
        import cv2
        cap = cv2.VideoCapture(video_path)
        if not cap.isOpened():
            return None
        # Try middle frame
        frame_count = int(cap.get(cv2.CAP_PROP_FRAME_COUNT) or 0)
        if frame_count > 0:
            cap.set(cv2.CAP_PROP_POS_FRAMES, max(0, frame_count // 2))
        ok, frame = cap.read()
        if not ok:
            cap.release()
            return None
        height, width = frame.shape[:2]
        # Convert BGR to RGB
        frame_rgb = frame[:, :, ::-1]
        img = Image.fromarray(frame_rgb)
        save_thumbnail(img, thumb_path, max_size=max_size)
        cap.release()
        return width, height
    except Exception:
        return None


def list_media_files(root_dir: str) -> List[str]:
    root_dir = os.path.abspath(root_dir)
    files: List[str] = []
    for current_root, _, fnames in os.walk(root_dir):
        for fname in fnames:
            fpath = os.path.join(current_root, fname)
            ext = os.path.splitext(fpath)[1].lower()
            if (
                (ext in SUPPORTED_IMAGE_EXTENSIONS)
                or (ext in SUPPORTED_VIDEO_EXTENSIONS)
                or (ext in SUPPORTED_AUDIO_EXTENSIONS)
            ):
                files.append(fpath)
    return files


def scan_directory(
    root_dir: str,
    db: Database,
    thumbnails_dir: Optional[str] = None,
    on_progress: Optional[Callable[[int, int, str], None]] = None,
) -> ScanResult:
    root_dir = os.path.abspath(root_dir)
    if thumbnails_dir is None:
        thumbnails_dir = os.path.join(db.base_dir, "thumbnails")
    ensure_dir(thumbnails_dir)

    files = list_media_files(root_dir)
    try:
        # Mark records for files that no longer exist as deleted
        db.mark_missing_files_deleted(files, root_dir)
    except Exception:
        pass
    total = len(files)

    scanned = 0
    added_or_updated = 0
    errors = 0

    # Load existing map for incremental check
    existing = db.existing_media_map_for_root(root_dir)

    for idx, fpath in enumerate(files, start=1):
        if on_progress is not None:
            try:
                on_progress(idx, total, fpath)
            except Exception:
                pass
        fname = os.path.basename(fpath)
        ext = os.path.splitext(fpath)[1].lower()
        try:
            stat = os.stat(fpath)
            size_bytes = int(stat.st_size)
            modified_time_utc = int(stat.st_mtime)

            # Incremental skip: if existing record matches size and mtime and has thumbnail, skip heavy work.
            prev = existing.get(fpath)
            if prev and prev.get("size_bytes") == size_bytes and prev.get("modified_time_utc") == modified_time_utc:
                thumb_path_existing = prev.get("thumbnail_path")
                sha256_existing = prev.get("sha256")
                if sha256_existing:
                    thumb_path = thumbnail_path_for(sha256_existing, thumbnails_dir)
                    if not os.path.exists(thumb_path):
                        # Create missing thumbnail based on media type
                        if ext in SUPPORTED_IMAGE_EXTENSIONS:
                            try:
                                with Image.open(fpath) as img:
                                    save_thumbnail(img, thumb_path)
                            except Exception:
                                thumb_path = None
                        elif ext in SUPPORTED_VIDEO_EXTENSIONS:
                            dims = create_video_thumbnail(fpath, thumb_path)
                            if dims is None:
                                thumb_path = None
                        else:
                            # Audio: no thumbnail generation
                            thumb_path = None
                    # If we created a new thumbnail, persist it
                    if thumb_path and thumb_path != thumb_path_existing:
                        try:
                            db.update_thumbnail_path(fpath, thumb_path)
                        except Exception:
                            pass
                # Nothing else to do for unchanged files
                scanned += 1
                continue

            # Changed or new file: compute sha and full metadata
            sha256 = sha256_of_file(fpath)

            width: Optional[int] = None
            height: Optional[int] = None
            p_hash_hex: Optional[str] = None
            captured_time = None
            thumb_path = thumbnail_path_for(sha256, thumbnails_dir)
            if ext in SUPPORTED_IMAGE_EXTENSIONS:
                media_type = "image"
            elif ext in SUPPORTED_VIDEO_EXTENSIONS:
                media_type = "video"
            elif ext in SUPPORTED_AUDIO_EXTENSIONS:
                media_type = "audio"
            else:
                media_type = "unknown"

            if ext in SUPPORTED_IMAGE_EXTENSIONS:
                with Image.open(fpath) as img:
                    width, height = img.size
                    try:
                        ph = imagehash.phash(img)
                        p_hash_hex = ph.__str__()
                    except Exception:
                        p_hash_hex = None
                    captured_time = exif_capture_time_utc(img)
                    if not os.path.exists(thumb_path):
                        try:
                            save_thumbnail(img, thumb_path)
                        except Exception:
                            thumb_path = None
            elif ext in SUPPORTED_VIDEO_EXTENSIONS:
                if not os.path.exists(thumb_path):
                    dims = create_video_thumbnail(fpath, thumb_path)
                    if dims is None:
                        thumb_path = None
                    else:
                        width, height = dims
            else:
                # Audio: do not attempt thumbnail; metadata remains minimal
                thumb_path = None

            media_id = db.upsert_media(
                file_path=fpath,
                root_dir=root_dir,
                file_name=fname,
                sha256=sha256,
                p_hash=p_hash_hex,
                width=width,
                height=height,
                size_bytes=size_bytes,
                captured_time_utc=captured_time,
                modified_time_utc=modified_time_utc,
                media_type=media_type,
                thumbnail_path=thumb_path,
            )
            if media_id:
                added_or_updated += 1
        except Exception as e:
            try:
                db.upsert_media(
                    file_path=fpath,
                    root_dir=root_dir,
                    file_name=fname,
                    sha256=None,
                    p_hash=None,
                    width=None,
                    height=None,
                    size_bytes=None,
                    captured_time_utc=None,
                    modified_time_utc=None,
                    media_type=(
                        "video" if is_video_file(fpath) else ("audio" if is_audio_file(fpath) else "image")
                    ),
                    thumbnail_path=None,
                    error=str(e),
                )
            except Exception:
                pass
            errors += 1
        finally:
            scanned += 1

    return ScanResult(scanned=scanned, added_or_updated=added_or_updated, errors=errors)

File: cpp/src/core/Scanner.cpp
#include "Scanner.h"
#include "Database.h"
#include "MediaRecord.h"

#include <QDir>
#include <QDirIterator>
#include <QFileInfo>
#include <QCryptographicHash>
#include <QImage>
#include <QImageReader>
#include <QPainter>
#include <QDebug>

#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/videoio.hpp>

namespace KeyTagger {

static const QSet<QString> IMAGE_EXTENSIONS = {
    ".jpg", ".jpeg", ".png", ".webp", ".bmp", ".tif", ".tiff", ".gif"
};

static const QSet<QString> VIDEO_EXTENSIONS = {
    ".mp4", ".mov", ".avi", ".mkv", ".webm", ".m4v", ".wmv", ".3gp"
};

static const QSet<QString> AUDIO_EXTENSIONS = {
    ".m4a", ".mp3", ".wav", ".flac", ".ogg", ".aac"
};

// ======================== ScannerWorker ========================

ScannerWorker::ScannerWorker(Database* db, const QString& rootDir, 
                             const QString& thumbnailsDir, QObject* parent)
    : QObject(parent)
    , m_db(db)
    , m_rootDir(QDir(rootDir).absolutePath())
    , m_thumbnailsDir(thumbnailsDir)
{
    if (m_thumbnailsDir.isEmpty()) {
        m_thumbnailsDir = QDir(m_rootDir).filePath("thumbnails");
    }
}

void ScannerWorker::cancel() {
    m_cancelled = true;
}

QString ScannerWorker::computeSha256(const QString& filePath) {
    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly)) {
        return QString();
    }
    
    QCryptographicHash hash(QCryptographicHash::Sha256);
    
    const qint64 chunkSize = 1024 * 1024; // 1MB chunks
    while (!file.atEnd()) {
        hash.addData(file.read(chunkSize));
    }
    
    return hash.result().toHex();
}

QString ScannerWorker::computeImagePHash(const QString& filePath) {
    // Simple perceptual hash using DCT approach
    try {
        cv::Mat img = cv::imread(filePath.toStdString(), cv::IMREAD_GRAYSCALE);
        if (img.empty()) return QString();
        
        // Resize to 32x32
        cv::Mat resized;
        cv::resize(img, resized, cv::Size(32, 32), 0, 0, cv::INTER_LINEAR);
        
        // Convert to float
        cv::Mat floatImg;
        resized.convertTo(floatImg, CV_32F);
        
        // Apply DCT
        cv::Mat dct;
        cv::dct(floatImg, dct);
        
        // Take top-left 8x8
        cv::Mat dctLow = dct(cv::Rect(0, 0, 8, 8));
        
        // Compute mean (excluding DC component)
        double sum = 0;
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                if (i == 0 && j == 0) continue;
                sum += dctLow.at<float>(i, j);
            }
        }
        double mean = sum / 63.0;
        
        // Generate hash
        quint64 hash = 0;
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                if (dctLow.at<float>(i, j) > mean) {
                    hash |= (1ULL << (i * 8 + j));
                }
            }
        }
        
        return QString::number(hash, 16).rightJustified(16, '0');
    } catch (...) {
        return QString();
    }
}

bool ScannerWorker::createImageThumbnail(const QString& sourcePath, const QString& destPath, int maxSize) {
    try {
        QImage img(sourcePath);
        if (img.isNull()) return false;
        
        // Scale to fit in maxSize x maxSize
        QImage scaled = img.scaled(maxSize, maxSize, Qt::KeepAspectRatio, Qt::SmoothTransformation);
        
        // Convert to RGB if necessary
        if (scaled.hasAlphaChannel()) {
            QImage rgb(scaled.size(), QImage::Format_RGB32);
            rgb.fill(Qt::black);
            QPainter painter(&rgb);
            painter.drawImage(0, 0, scaled);
            painter.end();
            scaled = rgb;
        }
        
        QDir().mkpath(QFileInfo(destPath).absolutePath());
        return scaled.save(destPath, "JPEG", 85);
    } catch (...) {
        return false;
    }
}

bool ScannerWorker::createVideoThumbnail(const QString& sourcePath, const QString& destPath, int maxSize) {
    try {
        cv::VideoCapture cap(sourcePath.toStdString());
        if (!cap.isOpened()) return false;
        
        // Seek to middle frame
        int frameCount = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_COUNT));
        if (frameCount > 0) {
            cap.set(cv::CAP_PROP_POS_FRAMES, frameCount / 2);
        }
        
        cv::Mat frame;
        if (!cap.read(frame)) {
            cap.release();
            return false;
        }
        
        // Convert BGR to RGB
        cv::Mat rgb;
        cv::cvtColor(frame, rgb, cv::COLOR_BGR2RGB);
        
        // Create QImage
        QImage img(rgb.data, rgb.cols, rgb.rows, static_cast<int>(rgb.step), QImage::Format_RGB888);
        QImage copy = img.copy(); // Make a deep copy since cv::Mat will be destroyed
        
        cap.release();
        
        // Scale
        QImage scaled = copy.scaled(maxSize, maxSize, Qt::KeepAspectRatio, Qt::SmoothTransformation);
        
        QDir().mkpath(QFileInfo(destPath).absolutePath());
        return scaled.save(destPath, "JPEG", 85);
    } catch (...) {
        return false;
    }
}

QPair<int, int> ScannerWorker::getImageDimensions(const QString& filePath) {
    QImageReader reader(filePath);
    QSize size = reader.size();
    return {size.width(), size.height()};
}

QPair<int, int> ScannerWorker::getVideoDimensions(const QString& filePath) {
    try {
        cv::VideoCapture cap(filePath.toStdString());
        if (!cap.isOpened()) return {0, 0};
        
        int width = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH));
        int height = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_HEIGHT));
        cap.release();
        
        return {width, height};
    } catch (...) {
        return {0, 0};
    }
}

qint64 ScannerWorker::getImageCaptureTime(const QString& filePath) {
    // Try to extract EXIF DateTimeOriginal
    // This is a simplified version - for full EXIF support, consider using libexiv2
    QImageReader reader(filePath);
    QString text = reader.text("DateTimeOriginal");
    if (text.isEmpty()) {
        text = reader.text("DateTime");
    }
    
    if (!text.isEmpty()) {
        // Parse "YYYY:MM:DD HH:MM:SS" format
        QDateTime dt = QDateTime::fromString(text, "yyyy:MM:dd HH:mm:ss");
        if (dt.isValid()) {
            return dt.toSecsSinceEpoch();
        }
    }
    
    return 0;
}

void ScannerWorker::process() {
    ScanResult result;
    
    // Get list of media files
    QStringList files = Scanner::listMediaFiles(m_rootDir);
    
    // Mark missing files as deleted
    try {
        m_db->markMissingFilesDeleted(files, m_rootDir);
    } catch (...) {}
    
    int total = files.size();
    
    // Get existing media map for incremental scanning
    auto existingMap = m_db->existingMediaMapForRoot(m_rootDir);
    
    QDir().mkpath(m_thumbnailsDir);
    
    for (int idx = 0; idx < total && !m_cancelled; ++idx) {
        const QString& filePath = files[idx];
        emit progress(idx + 1, total, filePath);
        
        QFileInfo fi(filePath);
        QString fileName = fi.fileName();
        QString ext = fi.suffix().toLower();
        ext = "." + ext;
        
        try {
            qint64 sizeBytes = fi.size();
            qint64 modifiedTimeUtc = fi.lastModified().toSecsSinceEpoch();
            
            // Check if we can skip this file
            if (existingMap.contains(filePath)) {
                auto& prev = existingMap[filePath];
                if (prev["size_bytes"].toLongLong() == sizeBytes &&
                    prev["modified_time_utc"].toLongLong() == modifiedTimeUtc &&
                    !prev["sha256"].toString().isEmpty()) {
                    
                    // Check if thumbnail exists
                    QString existingThumb = prev["thumbnail_path"].toString();
                    if (!existingThumb.isEmpty() && QFile::exists(existingThumb)) {
                        result.scanned++;
                        continue;
                    }
                    
                    // Create missing thumbnail
                    QString sha256 = prev["sha256"].toString();
                    QString thumbPath = QDir(m_thumbnailsDir).filePath(sha256 + ".jpg");
                    
                    bool thumbCreated = false;
                    if (Scanner::isImageFile(filePath)) {
                        thumbCreated = createImageThumbnail(filePath, thumbPath);
                    } else if (Scanner::isVideoFile(filePath)) {
                        thumbCreated = createVideoThumbnail(filePath, thumbPath);
                    }
                    
                    if (thumbCreated && thumbPath != existingThumb) {
                        m_db->updateThumbnailPath(filePath, thumbPath);
                    }
                    
                    result.scanned++;
                    continue;
                }
            }
            
            // Full processing for new or changed files
            QString sha256 = computeSha256(filePath);
            QString pHash;
            int width = 0, height = 0;
            qint64 capturedTime = 0;
            QString thumbPath = QDir(m_thumbnailsDir).filePath(sha256 + ".jpg");
            
            MediaType mediaType = MediaRecord::typeFromExtension(ext);
            
            if (mediaType == MediaType::Image) {
                pHash = computeImagePHash(filePath);
                auto dims = getImageDimensions(filePath);
                width = dims.first;
                height = dims.second;
                capturedTime = getImageCaptureTime(filePath);
                
                if (!QFile::exists(thumbPath)) {
                    createImageThumbnail(filePath, thumbPath);
                }
            } else if (mediaType == MediaType::Video) {
                auto dims = getVideoDimensions(filePath);
                width = dims.first;
                height = dims.second;
                
                if (!QFile::exists(thumbPath)) {
                    if (!createVideoThumbnail(filePath, thumbPath)) {
                        thumbPath.clear();
                    }
                }
            } else {
                // Audio - no thumbnail
                thumbPath.clear();
            }
            
            MediaRecord record;
            record.filePath = filePath;
            record.rootDir = m_rootDir;
            record.fileName = fileName;
            record.sha256 = sha256;
            record.pHash = pHash;
            if (width > 0) record.width = width;
            if (height > 0) record.height = height;
            record.sizeBytes = sizeBytes;
            if (capturedTime > 0) record.capturedTimeUtc = capturedTime;
            record.modifiedTimeUtc = modifiedTimeUtc;
            record.mediaType = mediaType;
            record.thumbnailPath = thumbPath;
            
            qint64 mediaId = m_db->upsertMedia(record);
            if (mediaId > 0) {
                result.addedOrUpdated++;
            }
            
        } catch (const std::exception& e) {
            qWarning() << "Error processing" << filePath << ":" << e.what();
            
            // Insert error record
            MediaRecord errorRecord;
            errorRecord.filePath = filePath;
            errorRecord.rootDir = m_rootDir;
            errorRecord.fileName = fileName;
            errorRecord.mediaType = MediaRecord::typeFromExtension(ext);
            errorRecord.error = QString::fromStdString(e.what());
            m_db->upsertMedia(errorRecord);
            
            result.errors++;
        }
        
        result.scanned++;
    }
    
    emit finished(result);
}

// ======================== Scanner ========================

Scanner::Scanner(Database* db, QObject* parent)
    : QObject(parent)
    , m_db(db)
{
}

Scanner::~Scanner() {
    cancel();
}

void Scanner::scanDirectory(const QString& rootDir, const QString& thumbnailsDir) {
    if (isRunning()) {
        cancel();
    }
    
    m_workerThread = new QThread(this);
    m_worker = new ScannerWorker(m_db, rootDir, thumbnailsDir);
    m_worker->moveToThread(m_workerThread);
    
    connect(m_workerThread, &QThread::started, m_worker, &ScannerWorker::process);
    connect(m_worker, &ScannerWorker::progress, this, &Scanner::scanProgress);
    connect(m_worker, &ScannerWorker::finished, this, [this](ScanResult result) {
        emit scanFinished(result);
        m_workerThread->quit();
    });
    connect(m_worker, &ScannerWorker::error, this, &Scanner::scanError);
    connect(m_workerThread, &QThread::finished, m_worker, &QObject::deleteLater);
    connect(m_workerThread, &QThread::finished, m_workerThread, &QObject::deleteLater);
    connect(m_workerThread, &QThread::finished, this, [this]() {
        m_worker = nullptr;
        m_workerThread = nullptr;
    });
    
    m_workerThread->start();
}

void Scanner::cancel() {
    if (m_worker) {
        m_worker->cancel();
    }
    if (m_workerThread) {
        m_workerThread->quit();
        m_workerThread->wait(5000);
    }
}

bool Scanner::isRunning() const {
    return m_workerThread && m_workerThread->isRunning();
}

QStringList Scanner::listMediaFiles(const QString& rootDir) {
    QStringList files;
    QString absRoot = QDir(rootDir).absolutePath();
    
    QDirIterator it(absRoot, QDir::Files, QDirIterator::Subdirectories);
    while (it.hasNext()) {
        QString path = it.next();
        QString ext = QFileInfo(path).suffix().toLower();
        ext = "." + ext;
        
        if (IMAGE_EXTENSIONS.contains(ext) || 
            VIDEO_EXTENSIONS.contains(ext) || 
            AUDIO_EXTENSIONS.contains(ext)) {
            files.append(path);
        }
    }
    
    return files;
}

bool Scanner::isImageFile(const QString& path) {
    QString ext = "." + QFileInfo(path).suffix().toLower();
    return IMAGE_EXTENSIONS.contains(ext);
}

bool Scanner::isVideoFile(const QString& path) {
    QString ext = "." + QFileInfo(path).suffix().toLower();
    return VIDEO_EXTENSIONS.contains(ext);
}

bool Scanner::isAudioFile(const QString& path) {
    QString ext = "." + QFileInfo(path).suffix().toLower();
    return AUDIO_EXTENSIONS.contains(ext);
}

} // namespace KeyTagger

